### PR TITLE
Fix tuple + subcommand edge cases

### PR DESCRIPTION
**Pull request by brentyi**
Fixed some issues identified by @wookayin in #89!

Specifically:

1. We previously assumed that field names would always be unique. This was causing problems when we used `tyro.conf.arg(name="")` to erase the name of multiple fields, which can help collapse nested structures.
        - This syntax is kind of just a hack, probably should avoid recommending this and instead stick to `tyro.conf.OmitArgPrefixes`, `tyro.conf.OmitSubcommandPrefixes` etc.
2. Fixed an edge case with nesting union types inside of tuples. Tuples are treated a bit specially because they can either be collapsed into a single argument (`tuple[int, int, int]`) or expanded into many (`tuple[Dataclass1, Dataclass2, Dataclass3]`).

We've also:

1. Expanded the scope of `tyro.conf.OmitSubcommandPrefixes[]` a little bit. This is technically a breaking change, but it's a niche feature, and after checking GitHub I confirmed that it doesn't break any existing (public) systems.
2. Added the cases from #89 as tests.
3. Improved the automatic test generation for Python 3.11 testing (which uses new-style syntax, like `A | B` instead of `Union[A, B]`).

### CODE REVIEW

**Comment by wookayin** [comment]
Thanks. I will try it out and leave some reviews. Note that PEP-604 requires python 3.10, so need to bump minimum python versions. Or PEP563 (python 3.8+).

File: tests/test_py311_generated/test_forward_ref_generated.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 @dataclasses.dataclass
 class A1:
     x: int
-    bc: "Union[B, C]"
+    bc: "(B| C)"
```
Comment (wookayin) [inline tests/test_py311_generated/test_forward_ref_generated.py:12]: nit: B | C

File: tests/test_py311_generated/test_tuple_with_subcommands_generated.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+import tyro.conf
+
+T = TypeVar("T")
+
+
+@dataclasses.dataclass
+class Checkout(Generic[T]):
+    """Check out a branch."""
+
+    branch: str
```
Comment (wookayin) [inline tests/test_py311_generated/test_tuple_with_subcommands_generated.py:24]: Did you mean `T` here?

File: tests/test_py311_generated/test_tuple_with_subcommands_generated.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                tyro.conf.arg(name=""),
+            ],
+            Annotated[
+                (
+                    Annotated[Checkout[str], tyro.conf.subcommand(name="checkout")]
+                    | Annotated[Commit, tyro.conf.subcommand(name="commit")]
+                ),
+                tyro.conf.arg(name=""),
+            ]
+            # ], args=["--help"])
```
Comment (wookayin) [inline tests/test_py311_generated/test_tuple_with_subcommands_generated.py:78]: Please feel free to remove this line.

**Review by wookayin — COMMENTED** [review]
This looks really cool, what a huge improvement! 👍️👍️

File: tests/test_py311_generated/test_tuple_with_subcommands_generated.py (hunk 2 above)
Comment (brentyi) [reply tests/test_py311_generated/test_tuple_with_subcommands_generated.py]: yes, thanks!

File: tests/test_py311_generated/test_forward_ref_generated.py (hunk 1 above)
Comment (brentyi) [reply tests/test_py311_generated/test_forward_ref_generated.py]: done!

**Review by wookayin — APPROVED** [review]
Looks great, it fixes all the bugs on my use cases so far. Thanks!

**Pull request merged**